File: signaling_server.py
#!/usr/bin/env python3
"""
A simple WebSocket signaling server for WebRTC.

- Allows multiple clients to connect.
- Broadcasts received messages to all OTHER connected clients.
- Provides verbose logging on connect, disconnect, and message forwarding.

Usage:
    python3 signaling_server.py

Then open ws://localhost:8765 in your clients.
"""

import asyncio
import websockets
import json

clients = set()

async def handler(websocket, path):
    """
    Each new client connects here. We'll log connections, handle messages,
    and broadcast them to all other connected clients.
    """
    print(f"[INFO] New client connected: {websocket.remote_address}")
    clients.add(websocket)

    try:
        async for message in websocket:
            print(f"[DEBUG] Received message from {websocket.remote_address}: {message}")
            # Broadcast to all *other* clients
            for client in clients:
                if client != websocket:
                    print(f"[DEBUG] Forwarding message to {client.remote_address}")
                    await client.send(message)
    except websockets.ConnectionClosed as exc:
        print(f"[WARN] Connection closed for {websocket.remote_address}: {exc}")
    finally:
        clients.remove(websocket)
        print(f"[INFO] Client disconnected: {websocket.remote_address}")

async def main():
    host = "0.0.0.0"
    port = 8765
    print(f"[INFO] Starting signaling server on {host}:{port}...")

    async with websockets.serve(handler, host, port):
        print(f"[INFO] Signaling server running at ws://{host}:{port}")
        await asyncio.Future()  # Run forever

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("[INFO] Server stopped via KeyboardInterrupt.")

File: signaling_server.js
const WebSocket = require('ws');

const PORT = 8765;
const wss = new WebSocket.Server({ port: PORT });
const clients = new Map(); // Store clients with stream_id

wss.on('connection', (ws, req) => {
    const clientAddress = req.socket.remoteAddress;
    console.log(`[INFO] New client connected: ${clientAddress}`);

    ws.on('message', (message) => {
        console.log(`[DEBUG] Received message from ${clientAddress}: ${message}`);
        let parsedMsg;
        try {
            parsedMsg = JSON.parse(message);
        } catch (err) {
            console.error("[ERROR] Invalid JSON message:", message);
            return;
        }

        const streamId = parsedMsg.stream_id;
        if (!streamId) {
            console.error("[ERROR] Missing stream_id in message");
            return;
        }

        if (parsedMsg.type === "register") {
            // Register client with a stream_id
            clients.set(ws, streamId);
            console.log(`[INFO] Client ${clientAddress} registered for stream ${streamId}`);
            return;
        }

        // Forward message to other clients with the same stream_id
        clients.forEach((otherStreamId, client) => {
            if (client !== ws && client.readyState === WebSocket.OPEN && otherStreamId === streamId) {
                console.log(`[DEBUG] Forwarding message for stream ${streamId} to a client`);
                client.send(message);
            }
        });
    });

    ws.on('close', () => {
        clients.delete(ws);
        console.log(`[INFO] Client disconnected: ${clientAddress}`);
    });

    ws.on('error', (err) => {
        console.error(`[ERROR] Client error from ${clientAddress}: ${err}`);
    });
});

console.log(`Signaling server started on port ${PORT}`);
